### PR TITLE
Memoize Literal hashCode

### PR DESCRIPTION
**Pull request by johnynek**
Literal.hashCode is coming up in a profiling code that uses Dagon. Since Binary nodes that actually merge the graph can cause duplicate work, hashCode can become exponential in the size of the graphs in pathological cases.

The ast you are optimizing should probably also consider similar optimizations since Dag calls these hashCode and equals quite a lot.

### CODE REVIEW

**Comment by johnynek** [comment]
cc @non @erik-stripe

File: core/src/main/scala/com/stripe/dagon/Literal.scala [hunk 1]
```diff
@@ -10,15 +10,68 @@ sealed trait Literal[N[_], T] {
...
+  case class Const[N[_], T](override val evaluate: N[T]) extends Literal[N, T] {
+    // cache hashCode since we hit it so much.
+    override val hashCode = (getClass, evaluate).hashCode
 
-  case class Unary[N[_], T1, T2](arg: Literal[N, T1], fn: N[T1] => N[T2]) extends Literal[N, T2]
+    // We often will cache the N[T] -> Literal[N, T] mapping, so we
+    // deal with referentially equal things, which can give cheap equality
+    override def equals(that: Any) =
+      that match {
+        case thatAR: AnyRef if thatAR eq this => true
```
Comment (travisbrown-stripe) [inline core/src/main/scala/com/stripe/dagon/Literal.scala]: It seems like the reference identity check could be factored out into `Literal` to clean up this code a bit?

File: core/src/main/scala/com/stripe/dagon/Literal.scala [hunk 2]
```diff
@@ -10,15 +10,68 @@ sealed trait Literal[N[_], T] {
...
+        case Const(thatE) => evaluate == thatE
+        case _ => false
+      }
+  }
+
+  case class Unary[N[_], T1, T2](arg: Literal[N, T1], fn: N[T1] => N[T2]) extends Literal[N, T2] {
+    // cache hashCode since we hit it so much.
+    override val hashCode = (getClass, arg, fn).hashCode
```
Comment (travisbrown-stripe) [inline core/src/main/scala/com/stripe/dagon/Literal.scala]: Is there a reason to prefer this approach over `MurmurHash3.productHash(this)`, which I feel like I've seen more often?

File: core/src/main/scala/com/stripe/dagon/FunctionK.scala [hunk 3]
```diff
@@ -11,3 +11,10 @@ trait FunctionK[T[_], R[_]] {
 
   def toFunction[U]: T[U] => R[U]
 }
+
+object FunctionK {
```
Comment (travisbrown-stripe) [inline core/src/main/scala/com/stripe/dagon/FunctionK.scala]: Is this related?

**Comment by travisbrown-stripe** [comment]
A couple of questions but LGTM.

File: core/src/main/scala/com/stripe/dagon/Literal.scala (hunk 2 above)
Comment (oscar-stripe) [reply core/src/main/scala/com/stripe/dagon/Literal.scala]: the only difference is that I am hashing the full class and productHash just hashes the string of the prefix of the name.

Additionally, I avoid an import with this. If you really want the productHash, I can do it.

File: core/src/main/scala/com/stripe/dagon/FunctionK.scala (hunk 3 above)
Comment (oscar-stripe) [reply core/src/main/scala/com/stripe/dagon/FunctionK.scala]: nope, I thought I wanted it, but I realized I can't use it as I was hoping. Would you like me to remove it?

File: core/src/main/scala/com/stripe/dagon/Literal.scala (hunk 1 above)
Comment (oscar-stripe) [reply core/src/main/scala/com/stripe/dagon/Literal.scala]: How were you thinking? Doesn't seem like I can actually get shorter code, and I also worry slightly about worse preformance.
